egs/wsj0-mix/DeepClustering/model.py

```diff
@@ -0,0 +1,123 @@
+    """
+
+    def __init__(self, rnn_type, input_size, hidden_size, n_layers=1,
+                 dropout=0, bidirectional=False):
+        super(SingleRNN, self).__init__()
+        rnn_type = rnn_type.upper()
+        assert rnn_type.upper() in ["RNN", "LSTM", "GRU"]
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.rnn = getattr(nn, rnn_type)(input_size, hidden_size,
+                                         num_layers=n_layers,
+                                         dropout=dropout,
+                                         batch_first=True,
+                                         bidirectional=bool(bidirectional))
+        self.rnn_type = rnn_type
+
+    def forward(self, inp):
+        """ Input shape [batch, seq, feats] """
+        output = inp
+        rnn_output, _ = self.rnn(output)
+        return rnn_output
+
+
+class ChimeraPP(nn.Module):
+    """ Chimera plus plus model used in deep clustering.
+
+    Args:
+        rnn_type: string, select from `'RNN'`, `'LSTM'`, `'GRU'`. Can
+            also be passed in lowercase letters.
+        n_src: int > 0. Number of masks to estimate.
+        input_size: input dimension to the network a.k.a number of frequency bins
+        embedding_dim: int. Dimension of the embedding vector
+        n_layers: int. Number of rnn layers
+        hidden_size: int. Number of non-linear elements in the hidden layer
+        dropout: float. In range [0-1]
+        bidirectional: Boolen. Is your rnn bi-directional
+    """
+    def __init__(self, rnn_type, n_src, input_size, \
+            embedding_dim=20, n_layers=2, hidden_size=600, \
+            dropout=0, bidirectional=True):
+        super(ChimeraPP, self).__init__()
+        self.input_dim = input_size
+        self.n_src = n_src
+        self.embedding_dim = embedding_dim
+        self.rnn = SingleRNN(rnn_type, input_size, hidden_size, n_layers, \
+            dropout, bidirectional)
+        self.dropout = nn.Dropout(dropout)
+        rnn_out_dim = hidden_size * 2 if bidirectional else hidden_size
+        self.embedding_layer = nn.Linear(rnn_out_dim, \
+                input_size * embedding_dim)
+        self.mask_layer = nn.Linear(rnn_out_dim, input_size * n_src)
+        self.non_linearity = nn.Sigmoid()
+
+    def forward(self, input_data):
+        batches, freq_dim, seq_cnt = input_data.shape
+        out = self.rnn(input_data.permute(0,2,1))
+        out = self.dropout(out)
+        projection = self.embedding_layer(out)
+        projection = self.non_linearity(projection)
+        projection = projection.view(batches, -1, self.embedding_dim)
+        proj_norm = th.norm(projection, p=2, dim=-1, keepdim=True) + \
+                th.finfo(th.float32).eps
+        projection_final =  projection/proj_norm
+        mask_out = self.mask_layer(out)
+        mask_out = mask_out.view(batches, self.n_src, self.input_dim, seq_cnt)
+        return projection_final, mask_out
```

Comment:
May be better to organize the .masknn module in a more hierarchical way and not put everything into blocks.py

Reply:
I agree, it's not scalable in this way. You can always propose changes in a PR if you feel like ;-)

Reply:
It perhaps makes sense to differentiate between blocks and models. For example SingleRNN should be inside blocks since it is used across different models and Chimerapp, DPRNN et al are models which are unlikely to be used across difference egs. Should you put models in the core or inside egs is debatable.

Reply:
The models can be used in recipes on different datasets, they have to be globally importable, they belong to core IMO.

Reply:
Okay to leave at least some models on masknn but to differentiate models and blocks like SingleRNN I think is reasonable. we can have masknn.blocks module folder and then in masknn all the models for example.